### PR TITLE
Passthrough virtualization feature support

### PR DESCRIPTION
**Pull request by rajkumar-xilinx**
Add passthrough virtualization feature support in XRT

http://jira.xilinx.com/browse/PR-13362

### CODE REVIEW

**Comment by rajkumar-xilinx** [comment]
retest this please

File: src/runtime_src/core/pcie/driver/linux/xocl/mgmtpf/mgmt-core.c [hunk 1]
```diff
@@ -951,6 +951,30 @@ void xclmgmt_connect_notify(struct xclmgmt_dev *lro, bool online)
 	vfree(mb_req);
 }
 
+static int xocl_get_p2p_bar_addr(struct xclmgmt_dev *xdev,
```
Comment (houlz0507) [inline src/runtime_src/core/pcie/driver/linux/xocl/mgmtpf/mgmt-core.c]: 1. For mailbox protocol, mgmtpf side should not get information from user side. Instead, userpf driver should send p2p bar addr to mgmtpf side.
2. when xocl_p2p_mem_reserve() is called, that means p2p bar is remapped. userpf driver needs to send the new addr to mgmtpf side
3. when mgmtpf driver is up, it notifies the userpf driver (MAILBOX_REQ_MGMT_STATE).  userpf should send the p2p addr to mgmpf

Comment (rajkumar-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/mgmtpf/mgmt-core.c]: Hi Lizhi, I have taken care of your review comments. Please review again and let me know. THanks.

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drv.c [hunk 2]
```diff
@@ -59,6 +59,49 @@ static void xocl_mb_connect(struct xocl_dev *xdev);
...
+		/* We should have one context for ourselves. */
+		BUG_ON(list_empty(&xdev->ctx_list));
+		userpf_err(xdev, "device is in use, can't update p2p in mgmtpf\n");
+		ret = -EBUSY;
+	}
+	mutex_unlock(&xdev->dev_lock);
+	if (ret < 0)
+		return;
+
+	ret = xocl_peer_request(xdev, mb_req, sizeof(struct mailbox_req),
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drv.c]: This call should be  xocl_peer_request(xdev, mb_req, req_len, ... ). Please update it.

File: src/runtime_src/core/pcie/driver/linux/xocl/mgmtpf/mgmt-core.c [hunk 3]
```diff
@@ -825,6 +825,40 @@ void xclmgmt_mailbox_srv(void *arg, void *data, size_t len,
...
+	case MAILBOX_REQ_READ_P2P_BAR_ADDR: {
+		struct pci_dev *pdev = lro->pci_dev;
+		struct mailbox_p2p_bar_addr *mb_p2p =
+			(struct mailbox_p2p_bar_addr*)req->data;
+		resource_size_t p2p_bar_addr = 0, p2p_bar_len = 0, range = 0;
+		u32 p2p_addr_base, range_base, final_val;
+
+		/* Passthrough Virtualization feature configuration */
+		if (xocl_passthrough_virtualization_on(lro)) {
+			if (iommu_present(&pci_bus_type)) {
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/pcie/driver/linux/xocl/mgmtpf/mgmt-core.c]: This should be iommu_present(&pci_bus_type). Please make the change

Comment (chvamshi-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/mgmtpf/mgmt-core.c]: changes should be done only on iommu_preset false case

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drv.c [hunk 4]
```diff
@@ -393,6 +436,7 @@ static void xocl_mailbox_srv(void *arg, void *data, size_t len,
 			/* Mgmt is online, try to probe peer */
 			userpf_info(xdev, "mgmt driver online\n");
 			(void) xocl_mb_connect(xdev);
+			(void) xocl_mb_read_p2p_addr(xdev);
```
[the remark this answers is not in the document]
Comment (rajkumar-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drv.c]: Hi Lizhi, I have taken care of this suggestion. I moved the mb_read_p2p call to reserver() function itself now. So, p2p enable/disable cases will be handled.

**Pull request merged**